homeassistant/components/coinbase/config_flow.py

```diff
@@ -0,0 +1,200 @@
+from .const import (
+    API_ACCOUNT_CURRENCY,
+    API_RATES,
+    CONF_CURRENCIES,
+    CONF_EXCHANGE_RATES,
+    CONF_YAML_API_TOKEN,
+    DOMAIN,
+    RATES,
+    WALLETS,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+STEP_USER_DATA_SCHEMA = vol.Schema(
+    {
+        vol.Required(CONF_API_KEY): str,
+        vol.Required(CONF_API_TOKEN): str,
+    }
+)
+
+
+async def validate_api(hass: core.HomeAssistant, data):
+    """Validate the credentials."""
+
+    for entry in hass.config_entries.async_entries(DOMAIN):
+        if entry.data[CONF_API_KEY] == data[CONF_API_KEY]:
+            raise AlreadyConfigured
+    try:
+        client = await hass.async_add_executor_job(
+            Client, data[CONF_API_KEY], data[CONF_API_TOKEN]
+        )
+        user = await hass.async_add_executor_job(client.get_current_user)
+    except AuthenticationError as error:
+        raise InvalidAuth from error
+    except ConnectionError as error:
+        raise CannotConnect from error
+
+    return {"title": user["name"]}
+
+
+async def validate_options(
+    hass: core.HomeAssistant, config_entry: config_entries.ConfigEntry, options
+):
+    """Validate the requested resources are provided by API."""
+
+    client = hass.data[DOMAIN][config_entry.entry_id].client
+
+    accounts = get_accounts(client)
+
+    accounts_currencies = [account[API_ACCOUNT_CURRENCY] for account in accounts]
+    available_rates = await hass.async_add_executor_job(client.get_exchange_rates)
+    if CONF_CURRENCIES in options:
+        for currency in options[CONF_CURRENCIES]:
+            if currency not in accounts_currencies:
+                raise CurrencyUnavaliable
+
+    if CONF_EXCHANGE_RATES in options:
+        for rate in options[CONF_EXCHANGE_RATES]:
+            if rate not in available_rates[API_RATES]:
+                raise ExchangeRateUnavaliable
+
+    return True
+
+
+class ConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle a config flow for Coinbase."""
+
+    VERSION = 1
+    CONNECTION_CLASS = config_entries.CONN_CLASS_CLOUD_POLL
+
+    async def async_step_user(self, user_input=None):
+        """Handle the initial step."""
+        errors = {}
+        if user_input is None:
+            return self.async_show_form(
+                step_id="user", data_schema=STEP_USER_DATA_SCHEMA, errors=errors
+            )
+
+        try:
+            info = await validate_api(self.hass, user_input)
+        except AlreadyConfigured:
+            return self.async_abort(reason="already_configured")
+        except CannotConnect:
+            errors["base"] = "cannot_connect"
+        except InvalidAuth:
+            errors["base"] = "invalid_auth"
+        except Exception:  # pylint: disable=broad-except
+            _LOGGER.exception("Unexpected exception")
+            errors["base"] = "unknown"
+        else:
+            return self.async_create_entry(
+                title=info["title"],
+                data=user_input,
```

Comment:
This now causes `CONF_CURRENCIES` and `CONF_EXCHANGE_RATES` to end up in `data`. I think they belong in `options`.